File: bench/strbuild/strbuild_bench.pas
program strbuild_bench;
{$mode ObjFPC}{$H+}

uses
  SysUtils, Classes,
  nextpas.core.base,
  nextpas.core.time.base,
  nextpas.core.bench,
  nextpas.core.bench.intf,
  nextpas.core.text.conv,
  nextpas.core.text.builder;

const
  N = 100000;

var
  GParts: array[0..N-1] of string;
  I: Integer;

procedure InitData;
begin
  for I := 0 to N - 1 do
    GParts[I] := 'item_' + IntToStr(I);
end;

procedure BenchBuilderAppend(const ACtx: IBenchContext);
var
  B: TBufStringBuilder;
  J: Integer;
begin
  B.Init(N * 16);
  for J := 0 to N - 1 do
  begin
    B.AppendStr(GParts[J]);
    B.AppendChar(',');
  end;
  ACtx.SetBytes(B.Len);
  B.Done;
end;

procedure BenchBuilderIntAppend(const ACtx: IBenchContext);
var
  B: TBufStringBuilder;
  J: Integer;
begin
  B.Init(N * 16);
  for J := 0 to N - 1 do
  begin
    B.AppendInt(J);
    B.AppendChar(',');
  end;
  ACtx.SetBytes(B.Len);
  B.Done;
end;

procedure BenchConcat(const ACtx: IBenchContext);
var
  S: string;
  J: Integer;
begin
  S := '';
  for J := 0 to N - 1 do
    S := S + GParts[J] + ',';
  ACtx.SetBytes(Length(S));
  if S = '' then WriteLn('');
end;

procedure BenchBuilderLarge(const ACtx: IBenchContext);
var
  B: TBufStringBuilder;
  J: Integer;
begin
  B.Init(N * 32);
  for J := 0 to N - 1 do
  begin
    B.AppendStr('line_');
    B.AppendInt(J);
    B.AppendStr(': value=');
    B.AppendFloat(J * 3.14);
    B.AppendChar(#10);
  end;
  ACtx.SetBytes(B.Len);
  B.Done;
end;

var
  LSuite: IBenchSuite;
  LResults: IBenchResults;
begin
  InitData;

  WriteLn('=== nextPas String Builder Benchmark ===');
  WriteLn('N=', N, ' parts');
  WriteLn;

  LSuite := TBenchSuite.Create('StrBuild')
    .SetMinDuration(TDuration.FromMilliseconds(100))
    .SetMaxIterations(10000)
    .SetMinSamples(6)
    .SetWarmupIters(3);

  LSuite.Add('Builder/Append/100k', @BenchBuilderAppend);
  LSuite.Add('Builder/IntAppend/100k', @BenchBuilderIntAppend);
  LSuite.Add('Concat/100k', @BenchConcat);
  LSuite.Add('Builder/Large/100k', @BenchBuilderLarge);

  LResults := LSuite.Run;

  WriteLn;
  WriteLn('=== benchstat format ===');
  WriteLn(LResults.ToBenchStat);
end.

File: bench/strbuild/benches/strbuild_bench.rs
use criterion::{criterion_group, criterion_main, Criterion};
use std::fmt::Write;
use std::hint::black_box;

const N: usize = 100_000;

fn make_parts() -> Vec<String> {
    (0..N).map(|i| format!("item_{}", i)).collect()
}

fn bench_builder_append(c: &mut Criterion) {
    let parts = make_parts();
    c.bench_function("Builder/Append/100k", |b| {
        b.iter(|| {
            let mut sb = String::with_capacity(N * 16);
            for p in &parts {
                sb.push_str(p);
                sb.push(',');
            }
            black_box(sb.len());
        })
    });
}

fn bench_builder_int_append(c: &mut Criterion) {
    c.bench_function("Builder/IntAppend/100k", |b| {
        b.iter(|| {
            let mut sb = String::with_capacity(N * 16);
            for i in 0..N {
                write!(sb, "{}", i).unwrap();
                sb.push(',');
            }
            black_box(sb.len());
        })
    });
}

fn bench_concat(c: &mut Criterion) {
    let parts = make_parts();
    c.bench_function("Concat/100k", |b| {
        b.iter(|| {
            let mut s = String::new();
            for p in &parts {
                s += p;
                s.push(',');
            }
            black_box(s.len());
        })
    });
}

fn bench_builder_large(c: &mut Criterion) {
    c.bench_function("Builder/Large/100k", |b| {
        b.iter(|| {
            let mut sb = String::with_capacity(N * 32);
            for i in 0..N {
                write!(sb, "line_{}: value={}\n", i, i as f64 * 3.14).unwrap();
            }
            black_box(sb.len());
        })
    });
}

criterion_group!(benches, bench_builder_append, bench_builder_int_append, bench_concat, bench_builder_large);
criterion_main!(benches);
